src/norm_flows.jl

```diff
@@ -0,0 +1,139 @@
+using Distributions
+using LinearAlgebra
+using Random
+using StatsFuns: softplus
+using Roots, LinearAlgebra # for inverse
+
+################################################################################
+#                            Planar and Radial Flows                           #
+#             Ref: Variational Inference with Normalizing Flows,               #
+#               D. Rezende, S. Mohamed(2015) arXiv:1505.05770                  #
+################################################################################
+
+mutable struct PlanarLayer <: Bijector
+    w
+    u
+    u_hat
+    b
+end
+
+function get_u_hat(u, w)
+    # to preserve invertibility
+    u_hat = (
+        u + (planar_flow_m(transpose(w) * u) - transpose(w) * u)[1]
+        * w/(norm(w[:,1],2) ^ 2)
+        )
+end
+
+function update_u_hat!(flow::PlanarLayer)
+    flow.u_hat = get_u_hat(flow.u, flow.w)
+end
+
+
+function PlanarLayer(dims::Int, container=Array)
+    w = container(randn(dims, 1))
+    u = container(randn(dims, 1))
+    b = container(randn(1))
+    u_hat = get_u_hat(u, w)
+    return PlanarLayer(w, u, u_hat, b)
+end
+
+planar_flow_m(x) = -1 .+ softplus.(x) # for planar flow
+dtanh(x) = 1 .- (tanh.(x)).^2 # for planar flow
+
+function transform(flow::PlanarLayer, z)
+    return z + flow.u_hat*tanh.(transpose(flow.w)*z .+ flow.b)
+end
+
+function forward(flow::T, z) where {T<:PlanarLayer}
+    update_u_hat!(flow)
+    # Compute log_det_jacobian
+    psi = ψ(z, flow.w, flow.b)
+    log_det_jacobian = log.(abs.(1.0 .+ transpose(psi)*flow.u_hat))
+
+    return (rv=transformed, logabsdetjacob=log_det_jacobian)
+end
+
+function inv(flow::PlanarLayer, y)
+    function f(y)
+        return loss(alpha) = (
+                (transpose(flow.w)*y)[1] - alpha
+                - (transpose(flow.w)*flow.u_hat)[1]
+                * tanh(alpha+flow.b[1])
+            )
+    end
+    alphas_ = [find_zero(f(y[:,i:i]), 0.0, Order16()) for i in 1:size(y)[2]]
+    alphas = transpose(alphas)
+    z_para = (flow.w ./norm(flow.w,2)) * alphas
+    z_per = (
+            y - z_para - flow.u_hat*tanh.(
+                                    transpose(flow.w) * z_para
+                                    .+ flow.b
+                                    )
+            )
+
+    return z_para+z_per
+end
+
+mutable struct RadialLayer <: Bijector
+    α_
+    β
+    z_0
+end
+
+function RadialLayer(dims::Int, container=Array)
+    α_ = container(randn(1))
+    β = container(randn(1))
+    z_0 = container(randn(dims, 1))
+    return RadialLayer(α_, β, z_0)
+end
+
+
+
+ψ(z, w, b) = dtanh(transpose(w) * z .+ b) .* w # for planar flow
+h(α, r) = 1 ./ (α .+ r) # for radial flow
+dh(α, r) = -h(α, r).^2 # for radial flow
+
+
+function transform(flow::RadialLayer, z)
+    α = softplus(flow.α_[1])
+    β_hat = -α + softplus(flow.β[1])
+    r = transpose(norm.([z[:,i] .- flow.z_0[:,:] for i in 1:size(z)[2]], 1))
```

Comment:
@sharanry I think `norm(..., 1)` is not taking then norm of a column vector.

Reply:
I just tried it. Looks fine to me.
`[z[:,i] .- flow.z_0[:,:] for i in 1:size(z)[2]]`

> 10-element Array{Array{Float64,2},1}:
>  [1.95363; 0.794311]   
>  [0.92017; -0.331166]  
>  [2.21721; -0.672726]  
>  [0.353866; 1.43099]   
>  [-0.239269; -0.193987]
>  [1.7601; 1.63398]     
>  [0.0194596; -0.557152]
>  [-0.198301; -2.51616] 
>  [1.82486; 0.370962]   
>  [2.80735; -2.52755]

`norm.([z[:,i] .- flow.z_0[:,:] for i in 1:size(z)[2]], 1)`

> 10-element Array{Float64,1}:
>  2.7479455500013144
>  1.2513367114426783
>  2.8899319657071585
>  1.7848539327206359
>  0.4332558976309925
>  3.3940802993947514
>  0.5766113670620187
>  2.7144649019585425
>  2.1958241235974283
>  5.33490420392606

Reply:
What's the expected L2 norm of `[1.95363; 0.794311]`?

Reply:
2.108...
`norm.([z[:,i] .- flow.z_0[:,:] for i in 1:size(z)[2]], 2)`

> 10-element Array{Float64,1}:
>  2.1089377103381453
>  0.9779491632989848
>  2.317015619955912 
>  1.474092343761427 
>  0.3080269486202933
>  2.4016335503846764
>  0.5574915338721331
>  2.523966236956232 
>  1.8621854866754475
>  3.777531594815805

Reply:
This looks correct to me.